cou/steps/hypervisor.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2024 Canonical Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Hypervisor planner."""
+import logging
+from dataclasses import dataclass
+
+from cou.apps.base import ApplicationUnit, OpenStackApplication
+from cou.steps import (
+    HypervisorUpgradePlan,
+    PostUpgradeStep,
+    PreUpgradeStep,
+    UpgradePlan,
+)
+from cou.utils.openstack import OpenStackRelease
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class Hypervisor:
+    """Hypervisor containing units for multiple applications."""
+
+    name: str
+    units: list[ApplicationUnit]
+
+
+@dataclass(frozen=True)
+class AZ:
+    """Availability Zone hypervisors in the az."""
+
+    name: str
+    hypervisors: list[Hypervisor]
+
+
+def verify_apps(apps: list[OpenStackApplication]) -> None:
+    """Verify OpenStack applications.
+
+    This function will verify that all machines contain nova-compute application.
```

Comment:
I didn't understand what you are going to do here. Are you going to some how filter all the applications that are colocated with nova-compute?

Reply:
The name is verify, I just want to make sure that machine passing to the HypervisorUpgradePlanner via apps has nova-compute on it, so yes I will verify that apps are either nova-compute or colocated apps.

It's possible that this will change to filter function and we will use it during creation of HypervisorUpgradePlanner.

Reply:
I think we don't need to double check that the machines passed to the hypervisors are nova-compute, because this check should happen before when we are validating user input.

However I think we'll need to do other stuff, so I'm not blocking this.